### PR TITLE
Volume multi-attach

### PR DESCRIPTION
**Pull request by jackhaibo**
**What this PR does / why we need it**:
Multi-attach for sushi, attach volume more than once.
Please see design spec for details.
https://github.com/opensds/design-specs/pull/26

**Special notes for your reviewer**:
@leonwanghui @xing-yang @xxwjj please review. Thanks.

### CODE REVIEW

File: csi/server/plugin/opensds/constants.go [hunk 1]
```diff
@@ -33,6 +33,7 @@ const (
 	KVolumeLvPath        = "lvPath"
 	KVolumeReplicationId = "replicationId"
 	KVolumeFstype        = "fstype"
+	KMultiAttach         = "multiattach"
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/constants.go]: s/multiattach/multiAttach

File: csi/server/plugin/opensds/constants.go [hunk 2]
```diff
@@ -42,6 +43,7 @@ const (
 	KPublishAttachId          = "AttachmentId"
 	KPublishSecondaryAttachId = "SecondaryAttachmentId"
 	KPublishAttachStatus      = "AttachmentStatus"
+	KPublishAttachMode        = "attachmode"
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/constants.go]: s/attachmode/attachMode

File: vendor/github.com/opensds/opensds/pkg/db/drivers/etcd/etcd.go [hunk 3]
```diff
@@ -1520,10 +1520,12 @@ func (c *Client) UpdateVolumeAttachment(ctx *c.Context, attachmentId string, att
 	if attachment.Metadata != nil {
 		result.Metadata = utils.MergeStringMaps(result.Metadata, attachment.Metadata)
 	}
-	// Update onnectionData
-	for k, v := range attachment.ConnectionData {
-		result.ConnectionData[k] = v
-	}
+	// Update connectionData
```
Comment (leonwanghui) [inline vendor/github.com/opensds/opensds/pkg/db/drivers/etcd/etcd.go]: Don't update the vendor folder directly

File: csi/server/deploy/kubernetes/csi-nodeplugin-opensdsplugin.yaml [hunk 4]
```diff
@@ -14,6 +14,9 @@ spec:
       labels:
         app: csi-nodeplugin-opensdsplugin
     spec:
+      tolerations:
```
Comment (xing-yang) [inline csi/server/deploy/kubernetes/csi-nodeplugin-opensdsplugin.yaml]: Please add a note that this is only needed for multi-attach.

File: csi/server/plugin/opensds/controller.go [hunk 5]
```diff
@@ -82,31 +81,21 @@ func GetDefaultProfile() (*model.ProfileSpec, error) {
...
-func FindVolume(req *model.VolumeSpec) (bool, bool, *model.VolumeSpec, error) {
-	isExist := false
+func FindVolume(req *model.VolumeSpec) (*model.VolumeSpec, error) {
 	volumes, err := client.ListVolumes()
-
 	if err != nil {
-		glog.Error("List volumes failed: ", err)
-
-		return false, false, nil, err
+		msg := fmt.Sprintf("List volumes failed: %v", err)
```
Comment (xing-yang) [inline csi/server/plugin/opensds/controller.go]: error msg should start with lower case letter.

File: csi/server/plugin/opensds/controller.go [hunk 6]
```diff
@@ -131,7 +120,9 @@ func (p *Plugin) CreateVolume(
 	}
 
 	if client == nil {
-		return nil, status.Error(codes.InvalidArgument, "client is nil")
+		msg := "CreateVolume: the client is nil"
```
Comment (xing-yang) [inline csi/server/plugin/opensds/controller.go]: start msg with lower case letter.

File: csi/server/plugin/opensds/controller.go [hunk 7]
```diff
@@ -384,71 +385,72 @@ func isStringMapEqual(metadataA, metadataB map[string]string) bool {
...
+	volMultiAttach bool) error {
+
 	glog.V(5).Infof("start to isVolumePublished, canAtMultiNode = %v, attachReq = %v",
 		canAtMultiNode, attachReq)
 
 	attachments, err := client.ListVolumeAttachments()
 	if err != nil {
-		glog.V(5).Info("ListVolumeAttachments failed: " + err.Error())
-		return nil, status.Error(codes.FailedPrecondition, err.Error())
+		msg := fmt.Sprintf("ListVolumeAttachments failed: %v", err)
```
Comment (xing-yang) [inline csi/server/plugin/opensds/controller.go]: start msg with lower case letter.

File: csi/server/plugin/opensds/node.go [hunk 8]
```diff
@@ -77,34 +79,40 @@ func updateAttachment(mountpoint string, key string, attachment *model.VolumeAtt
...
+// getVolumeAndAttachmentByVolumeId Get volume and attachment by volumeId
 func getVolumeAndAttachmentByVolumeId(volId string) (*model.VolumeSpec, *model.VolumeAttachmentSpec, error) {
 	if r := getReplicationByVolume(volId); r != nil {
 		volId = r.Metadata[KAttachedVolumeId]
 	}
 
 	vol, err := client.GetVolume(volId)
 	if nil != err || nil == vol {
-		return nil, nil, status.Error(codes.NotFound, "Volume does not exist")
+		msg := fmt.Sprintf("Volume does not exist, %v", err)
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: start msg with lower case letter.

File: csi/server/plugin/opensds/node.go [hunk 9]
```diff
@@ -77,34 +79,40 @@ func updateAttachment(mountpoint string, key string, attachment *model.VolumeAtt
...
+		glog.Error(msg)
+		return nil, nil, status.Error(codes.NotFound, msg)
 	}
 
 	attachments, err := client.ListVolumeAttachments()
 	if nil != err {
-		return nil, nil, status.Error(codes.FailedPrecondition, "List volume attachments failed")
+		msg := fmt.Sprintf("List volume attachments failed, %v", err)
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: start msg with lower case letter.

File: csi/server/plugin/opensds/node.go [hunk 10]
```diff
@@ -118,53 +126,61 @@ func getVolumeAndAttachmentByVolumeId(volId string) (*model.VolumeSpec, *model.V
 		}
 	}
 
+	if attachment == nil {
+		msg := fmt.Sprintf("attachment is not exist")
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: does not exist

File: csi/server/plugin/opensds/node.go [hunk 11]
```diff
@@ -118,53 +126,61 @@ func getVolumeAndAttachmentByVolumeId(volId string) (*model.VolumeSpec, *model.V
...
-		attachment.Metadata[key] = strings.Join(modifyPaths, ";")
+		attachment.Metadata[key] = strings.Join(paths, ";")
 	}
 
 	if KStagingTargetPath == key {
 		volConnector := connector.NewConnector(attachment.DriverVolumeType)
 
 		if volConnector == nil {
-			return status.Error(codes.FailedPrecondition, fmt.Sprintf("Unsupport driverVolumeType: %s", attachment.DriverVolumeType))
+			msg := fmt.Sprintf("Unsupport driverVolumeType: %s", attachment.DriverVolumeType)
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: start msg with lower case letter.

File: csi/server/plugin/opensds/node.go [hunk 12]
```diff
@@ -203,15 +219,20 @@ func (p *Plugin) NodeStageVolume(
...
 
 	// Check REQUIRED field
-	glog.V(5).Info("start to NodeStageVolume, Volume_id: " + req.VolumeId + ", staging_target_path: " + req.StagingTargetPath)
+	glog.V(5).Info("start to NodeStageVolume, Volume_id: " + req.VolumeId +
+		", staging_target_path: " + req.StagingTargetPath)
 	defer glog.V(5).Info("end to NodeStageVolume")
 
 	if "" == req.VolumeId || "" == req.StagingTargetPath || nil == req.VolumeCapability {
-		return nil, status.Error(codes.InvalidArgument, "Volume_id/staging_target_path/volume_capability must be specified")
+		msg := "Volume_id/staging_target_path/volume_capability must be specified"
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: start msg with lower case letter.

File: csi/server/plugin/opensds/node.go [hunk 13]
```diff
@@ -203,15 +219,20 @@ func (p *Plugin) NodeStageVolume(
...
+		glog.Error(msg)
+		return nil, status.Error(codes.InvalidArgument, msg)
 	}
 
 	if client == nil {
-		return nil, status.Error(codes.InvalidArgument, "client is nil")
+		msg := "NodeStageVolume: the client is nil"
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: start msg with lower case letter.

File: csi/server/plugin/opensds/node.go [hunk 14]
```diff
@@ -346,15 +377,20 @@ func (p *Plugin) NodeUnstageVolume(
...
 
 	// Check REQUIRED field
-	glog.V(5).Info("start to NodeUnstageVolume, Volume_id: " + req.VolumeId + ", staging_target_path: " + req.StagingTargetPath)
+	glog.V(5).Info("start to NodeUnstageVolume, Volume_id: " + req.VolumeId +
+		", staging_target_path: " + req.StagingTargetPath)
 	defer glog.V(5).Info("end to NodeUnstageVolume")
 
 	if "" == req.VolumeId || "" == req.StagingTargetPath {
-		return nil, status.Error(codes.InvalidArgument, "Volume_id/staging_target_path must be specified")
+		msg := "Volume_id/staging_target_path must be specified"
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: start msg with lower case letter.

File: csi/server/plugin/opensds/node.go [hunk 15]
```diff
@@ -410,15 +447,20 @@ func (p *Plugin) NodePublishVolume(
...
 
 	// Check REQUIRED field
-	glog.V(5).Info("start to NodePublishVolume, Volume_id: " + req.VolumeId + ", staging_target_path: " + req.StagingTargetPath + ", target_path: " + req.TargetPath)
+	glog.V(5).Info("start to NodePublishVolume, Volume_id: " + req.VolumeId +
+		", staging_target_path: " + req.StagingTargetPath + ", target_path: " + req.TargetPath)
 	defer glog.V(5).Info("end to NodePublishVolume")
 
 	if "" == req.VolumeId || "" == req.StagingTargetPath || "" == req.TargetPath || nil == req.VolumeCapability {
-		return nil, status.Error(codes.InvalidArgument, "Volume_id/staging_target_path/target_path/volume_capability must be specified")
+		msg := "Volume_id/staging_target_path/target_path/volume_capability must be specified"
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: start msg with lower case letter.

File: csi/server/plugin/opensds/node.go [hunk 16]
```diff
@@ -510,11 +556,15 @@ func (p *Plugin) NodeUnpublishVolume(
 	defer glog.V(5).Info("end to NodeUnpublishVolume")
 
 	if "" == req.VolumeId || "" == req.TargetPath {
-		return nil, status.Error(codes.InvalidArgument, "Volume_id/target_path must be specified")
+		msg := "Volume_id/target_path must be specified"
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: start msg with lower case letter.

File: csi/server/plugin/opensds/constants.go (hunk 1 above)
Comment (jackhaibo) [reply csi/server/plugin/opensds/constants.go]: Please see the code below
```
for k, v := range req.GetParameters() {
		switch strings.ToLower(k) {
		case KVolumeFstype:
			fstype = v

		case KParamProfile:
			volumebody.ProfileId = v
		case KParamAZ:
			volumebody.AvailabilityZone = v
		case KParamEnableReplication:
			if strings.ToLower(v) == "true" {
				enableReplication = true
			}
		case KParamSecondaryAZ:
			secondaryAZ = v
		case KMultiAttach:
			glog.Info(KMultiAttach + ":" + strings.ToLower(v))
			if strings.ToLower(v) == "t" {
				volumebody.MultiAttach = true
			}
		case KPublishAttachMode:
			if strings.ToLower(v) == "ro" {
				attachMode = "ro"
			}
		}
	}
```

File: csi/server/plugin/opensds/constants.go (hunk 2 above)
Comment (jackhaibo) [reply csi/server/plugin/opensds/constants.go]: Same here.

File: csi/server/deploy/kubernetes/csi-nodeplugin-opensdsplugin.yaml (hunk 4 above)
Comment (jackhaibo) [reply csi/server/deploy/kubernetes/csi-nodeplugin-opensdsplugin.yaml]: Here means deploying `nodepod` on the master node.

File: csi/server/plugin/opensds/controller.go (hunk 5 above)
Comment (jackhaibo) [reply csi/server/plugin/opensds/controller.go]: Done

File: vendor/github.com/opensds/opensds/pkg/db/drivers/etcd/etcd.go (hunk 3 above)
Comment (jackhaibo) [reply vendor/github.com/opensds/opensds/pkg/db/drivers/etcd/etcd.go]: Done.

File: csi/server/plugin/opensds/controller.go [hunk 17]
```diff
@@ -384,90 +385,100 @@ func isStringMapEqual(metadataA, metadataB map[string]string) bool {
...
 	req *csi.ControllerPublishVolumeRequest) (
 	*csi.ControllerPublishVolumeResponse, error) {
 
 	glog.V(5).Info("start to ControllerPublishVolume")
 	defer glog.V(5).Info("end to ControllerPublishVolume")
 
 	if req.VolumeId == "" {
-		msg := "ControllerPublishVolume Volume ID must be provided"
+		msg := "controllerPublishVolume: Volume ID must be provided"
 		glog.Info(msg)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/controller.go]: glog.Error(msg)

File: csi/server/plugin/opensds/controller.go [hunk 18]
```diff
@@ -384,90 +385,100 @@ func isStringMapEqual(metadataA, metadataB map[string]string) bool {
...
 		return nil, status.Error(codes.InvalidArgument, msg)
 	}
 
 	if req.NodeId == "" {
-		msg := "ControllerPublishVolume Node ID must be provided"
+		msg := "controllerPublishVolume: Node ID must be provided"
 		glog.Info(msg)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/controller.go]: Same here.

File: csi/server/plugin/opensds/controller.go [hunk 19]
```diff
@@ -384,90 +385,100 @@ func isStringMapEqual(metadataA, metadataB map[string]string) bool {
...
 		return nil, status.Error(codes.InvalidArgument, msg)
 	}
 
 	if client == nil {
-		return nil, status.Error(codes.InvalidArgument, "client is nil")
+		msg := "controllerPublishVolume: the client is nil"
+		glog.Info(msg)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/controller.go]: Same here

File: csi/server/plugin/opensds/controller.go [hunk 20]
```diff
@@ -384,90 +385,100 @@ func isStringMapEqual(metadataA, metadataB map[string]string) bool {
...
+		msg := fmt.Sprintf("the volume %s is not exist, error info: %v",
+			req.VolumeId, err)
+		glog.Error(msg)
 		return nil, status.Error(codes.NotFound, msg)
 	}
 
 	pool, err := client.GetPool(volSpec.PoolId)
 	if err != nil || pool == nil {
-		msg := fmt.Sprintf("the pool %s is not sxist", volSpec.PoolId)
+		msg := fmt.Sprintf("the pool %s is not sxist, error info: %v",
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/controller.go]: s/sxist/exist

File: csi/server/plugin/opensds/controller.go [hunk 21]
```diff
@@ -525,64 +541,62 @@ func (p *Plugin) ControllerPublishVolume(
...
+			KMultiAttach:         attachMode,
 		},
 	}
 
 	if replicationId, ok := req.VolumeContext[KVolumeReplicationId]; ok {
 		r, err := client.GetReplication(replicationId)
 		if err != nil {
-			return nil, status.Error(codes.FailedPrecondition, "Get replication failed")
+			msg := fmt.Sprintf("failed to get replication: %v", err)
+			glog.Info(msg)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/controller.go]: glog.Error(msg)

File: csi/server/plugin/opensds/controller.go [hunk 22]
```diff
@@ -525,64 +541,62 @@ func (p *Plugin) ControllerPublishVolume(
...
 		}
 
 		attachReq.VolumeId = r.SecondaryVolumeId
-		existAttachment, err := isVolumePublished(canAtMultiNode, attachReq, expectedMetadata)
+
+		secondaryVolume, err := client.GetVolume(attachReq.VolumeId)
 		if err != nil {
-			return nil, err
+			msg := fmt.Sprintf("failed to get secondary volume: %v", err)
+			glog.Info(msg)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/controller.go]: Same here

File: csi/server/plugin/opensds/controller.go [hunk 23]
```diff
@@ -525,64 +541,62 @@ func (p *Plugin) ControllerPublishVolume(
...
+		if err != nil {
+			return nil, err
 		}
 
-		resp.PublishContext[KPublishSecondaryAttachId] = attachSpec.Id
+		newAttachment, errAttach := client.CreateVolumeAttachment(attachReq)
+		if errAttach != nil {
+			msg := fmt.Sprintf("failed to ControllerPublishVolume: the volume %s failed to publish to node %s, error info %v",
+				req.VolumeId, req.NodeId, errAttach)
+			glog.Errorf(msg)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/controller.go]: glog.Error(msg)

File: csi/server/plugin/opensds/node.go [hunk 24]
```diff
@@ -118,53 +126,61 @@ func getVolumeAndAttachmentByVolumeId(volId string) (*model.VolumeSpec, *model.V
...
-			return status.Error(codes.FailedPrecondition, fmt.Sprintf("Unsupport driverVolumeType: %s", attachment.DriverVolumeType))
+			msg := fmt.Sprintf("unsupport driverVolumeType: %s", attachment.DriverVolumeType)
+			glog.Error(msg)
+			return status.Error(codes.FailedPrecondition, msg)
 		}
 
 		err := volConnector.Detach(attachment.ConnectionData)
 		if err != nil {
-			return status.Errorf(codes.FailedPrecondition, "%s", err.Error())
+			msg := fmt.Sprintf("detach falied, %v", err)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/node.go]: s/falied/failed

File: csi/server/plugin/opensds/node.go [hunk 25]
```diff
@@ -332,7 +361,9 @@ func (p *Plugin) NodeStageVolume(
 
 	_, err = client.UpdateVolume(vol.Id, vol)
 	if err != nil {
-		return nil, status.Error(codes.FailedPrecondition, "update volume failed")
+		msg := fmt.Sprintf("update volume failed, %v", err)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/node.go]: Please change `,` to `:`

**Pull request closed without merging**